### PR TITLE
[https://github.com/eclipse/xtext-eclipse/issues/1221] Serializer.

### PR DESCRIPTION
**Pull request by miklossy**
- Improve the serializeReplacement implementation by modifying the
ReplaceRegion length calculation so that it takes not only the
ICompositeNode.getTotalLength() into account, but considers if the new
text contains additional whitespaces and the old text is also followed
by white spaces, than the whitespaces contained by the original document
is also consumed by the quickfix.

Signed-off-by: Tamas Miklossy <miklossy@itemis.de>

### CODE REVIEW

File: org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java [hunk 1]
```diff
@@ -210,7 +211,28 @@ public ReplaceRegion serializeReplacement(EObject obj, SaveOptions options) {
 			throw new IllegalStateException("Cannot replace an obj that has no associated node");
 		}
 		String text = serialize(obj, options);
-		return new ReplaceRegion(node.getTotalOffset(), node.getTotalLength(), text);
+		int replaceRegionLength = calculateReplaceRegionLength(node, text);
+		return new ReplaceRegion(node.getTotalOffset(), replaceRegionLength, text);
 	}
 
+	private int calculateReplaceRegionLength(ICompositeNode node, String text) {
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java:218]: Unfortunately this is not generally applicable, e.g. languages with synthetic tokens or semantic whitespace will break. This method must at least be protected

File: org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java [hunk 2]
```diff
@@ -210,7 +211,28 @@ public ReplaceRegion serializeReplacement(EObject obj, SaveOptions options) {
...
+			String remainingText = text.substring(oldTextLength);
+			if (containsOnlyWhiteSpaces(remainingText) && whiteSpaceFollows(node)) {
+				return newTextLength;
+			}
+		}
+		return oldTextLength;
+	}
+
+	private boolean whiteSpaceFollows(ICompositeNode node) {
+		return node.hasNextSibling() && node.getNextSibling() instanceof HiddenLeafNode;
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java:232]: Checks against implementation classes of the node model must be avoided. Please use `instanceof ILeafNode && node.isHidden()` instead.

File: org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java [hunk 3]
```diff
@@ -210,7 +211,28 @@ public ReplaceRegion serializeReplacement(EObject obj, SaveOptions options) {
...
+	}
+
+	private boolean containsOnlyWhiteSpaces(String text) {
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java:235]: Could be `CharMatcher.whitespac().matchesAllOf(text)`

File: org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java (hunk 2 above)
Comment (szarnekow) [reply org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java:232]: In any case, the logic here is unfortunately flawed. If I understand the intention behind this code correctly, it must be something like `node.getLeadNodes().head().isHidden()` - though the name `whiteSpaceFollows` is misleading given that comment nodes are also hidden.

File: org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java [hunk 4]
```diff
@@ -210,7 +211,28 @@ public ReplaceRegion serializeReplacement(EObject obj, SaveOptions options) {
...
+		if (newTextLength > oldTextLength) {
+			String remainingText = text.substring(oldTextLength);
+			if (containsOnlyWhiteSpaces(remainingText) && whiteSpaceFollows(node)) {
+				return newTextLength;
+			}
+		}
+		return oldTextLength;
+	}
+
+	private boolean whiteSpaceFollows(ICompositeNode node) {
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java:231]: whiteSpace -> whitespace

File: org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java (hunk 3 above)
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java:235]: containsOnlyWhiteSpaces -> isWhitespace

File: org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java (hunk 1 above)
Comment (miklossy) [reply org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java]: Fixed.

File: org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java (hunk 2 above)
Comment (miklossy) [reply org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java]: Fixed.

File: org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java (hunk 4 above)
Comment (miklossy) [reply org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java]: Fixed.

File: org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java (hunk 3 above)
Comment (miklossy) [reply org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java]: Fixed.

Comment (miklossy) [reply org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java]: Fixed.

File: org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java (hunk 1 above)
Comment (miklossy) [reply org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java]: > Unfortunately this is not generally applicable, e.g. languages with synthetic tokens or semantic whitespace will break.

Actually I think that it is really too dangerous to modify the code in the `xtext-core` code base. Hence it is quite difficult to provide a general solution there, I would prefer to fix the problem only on the Domainmodel example side. We can still consider a general solution later on, if lot of "client projects" claims about the problem, but for the time being a fix only for the Domainmodel example would satisfy my needs.

@cdietrich @szarnekow : If you agree, I will close this PR and provide another one in `xtext-eclipse`.

Comment (szarnekow) [reply org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java]: @miklossy I didn't want to scare you and I'm sure that you'll get this straight. If you don't mind, I'll be happy to give advice and provide feedback.

**Comment by miklossy** [comment]
@szarnekow I updated the PR as requested: I added a SerializerReplacementCalculationTest test case to the xtext-core repository based on the  [NoJdtTestLanguage](https://github.com/eclipse/xtext-core/blob/master/org.eclipse.xtext.testlanguages/src/org/eclipse/xtext/testlanguages/noJdt/NoJdtTestLanguage.xtext).

File: org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java [hunk 5]
```diff
@@ -210,7 +213,35 @@ public ReplaceRegion serializeReplacement(EObject obj, SaveOptions options) {
...
+				return newTextLength;
+			}
+		}
+		return oldTextLength;
+	}
+
+	protected boolean hiddenNodeFollows(ICompositeNode node) {
+		if (node.hasNextSibling()) {
+			INode nextSibling = node.getNextSibling();
+			if (nextSibling instanceof ILeafNode) {
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java:236]: If the next sibling is a composite node, it's first child may still be a hidden node.
If you want to catch those, too, I recommend to use something like nextSibling.getLeafNodes() and check the first node in that list.

**Comment by miklossy** [comment]
Thanks for the feedback @szarnekow . I updated the PR as requested.

> There may be corner cases where the next sibling does not exist but the next node is still a hidden node (if current is the last child in its parent).
The current fix is actually strong enough to catch two different use cases (`Domainmodel example` and `NoJdtTestLanguage`), so I considered this fix already as an improvement, even if it does not cover all use cases. I would consider the corner cases if we can find a use case that is not covered by the current fix, so we can translate that use case into a test case and improve the implementation based on that. But that I would consider as future work in a separate issue.

**Comment by szarnekow** [comment]
I assume that you are fully aware of the fact that Xtext is neither a framework that helps to run the domainmodel example nor a framework to build a painfully trivial test language.
If every bugfix was done under the proposition of ignoring corner cases, the framework would not be where it was now.

**Comment by miklossy** [comment]
> There may be corner cases where the next sibling does not exist but the next node is still a hidden node (if current is the last child in its parent).

Is there already any `getNextNode()` method that I can use for this purpose?

File: org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java [hunk 6]
```diff
@@ -210,7 +213,50 @@ public ReplaceRegion serializeReplacement(EObject obj, SaveOptions options) {
 			throw new IllegalStateException("Cannot replace an obj that has no associated node");
 		}
 		String text = serialize(obj, options);
-		return new ReplaceRegion(node.getTotalOffset(), node.getTotalLength(), text);
+		int replaceRegionLength = calculateReplaceRegionLength(node, text);
+		return new ReplaceRegion(node.getTotalOffset(), replaceRegionLength, text);
 	}
 
+	protected int calculateReplaceRegionLength(ICompositeNode node, String text) {
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java:220]: Please also add `@Since` tags to the new methods.

**Comment by miklossy** [comment]
`@since 2.22` tags have been added.

File: org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java [hunk 7]
```diff
@@ -210,7 +213,60 @@ public ReplaceRegion serializeReplacement(EObject obj, SaveOptions options) {
...
+				}	
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * @since 2.22
+	 */
+	private boolean isWhitespace(String text) {
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java:269]: Last thing that slipped through: This method should either be protected or the since tag is unnecessary.

Comment (miklossy) [reply org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java]: I changed the method visibility to `protected`.
